<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CDSeq: a complete deconvolution method using sequencing data • CDSeq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CDSeq: a complete deconvolution method using sequencing data">
<meta property="og:description" content="CDSeq">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CDSeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CDSeq-vignette.html">CDSeq: a complete deconvolution method using sequencing data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="CDSeq-vignette_files/header-attrs-2.6/header-attrs.js"></script><script src="CDSeq-vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CDSeq: a complete deconvolution method using sequencing data</h1>
            
      
      
      <div class="hidden name"><code>CDSeq-vignette.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>CDSeq is a complete deconvolution method using sequencing data. Simply put, CDSeq takes bulk RNA-Seq read counts data as input and estimates the cell-type-specific gene expression profiles (csGEPs) and sample-specific cell-type proportions (SSP) simultaneously.</p>
</div>
<div id="quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start</h1>
<p>The minimum inputs for CDSeq are: bulk RNAseq read counts, number of cell types in the mixtures (a <a href="#scalar">scalar</a> or a <a href="#vector">vector</a>), number of Markov chain Monte Carlo (MCMC) iterations. Set <em>cpu_number = 1</em> to avoid using more cpus in this simple example.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">result</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/CDSeq.html">CDSeq</a></span><span class="op">(</span>bulk_data <span class="op">=</span>  <span class="va">mixtureGEP</span>, 
              cell_type_number <span class="op">=</span> <span class="fl">6</span>, 
              mcmc_iterations <span class="op">=</span> <span class="fl">1000</span>, 
              cpu_number<span class="op">=</span><span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>Check out the details about <a href="#input">input</a> and <a href="#output">output</a> in the following.</p>
</div>
<div id="usage" class="section level1">
<h1 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h1>
<div id="input" class="section level2">
<h2 class="hasAnchor">
<a href="#input" class="anchor"></a>Input</h2>
<table class="table">
<thead><tr class="header">
<th align="center">Input</th>
<th align="center">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">bulk_data</td>
<td align="center">this is the bulk RNAseq read counts data. It is a G <span class="math inline">\(\times\)</span> M matrix or data frame, where G is the number of genes and M is the sample size.</td>
</tr>
<tr class="even">
<td align="center">beta</td>
<td align="center">hyperparameter for csGEP estimation.</td>
</tr>
<tr class="odd">
<td align="center">alpha</td>
<td align="center">hyperparameter for SSP estimation.</td>
</tr>
<tr class="even">
<td align="center">cell_type_number</td>
<td align="center">number of cell types. It is an integer or a vector of varying integers.</td>
</tr>
<tr class="odd">
<td align="center">mcmc_iterations</td>
<td align="center">number of iterations for Gibbs sampler.</td>
</tr>
<tr class="even">
<td align="center">dilution_factor</td>
<td align="center">dilution factor is a number used to dilute the input bulk data for faster computation. CDSeq will dilute the bulk data by <em>dilution_factor</em>, i.e. <span class="math inline">\(\frac{\text{bulk_data}}{\text{dilution_factor}}\)</span> will be analyzed.</td>
</tr>
<tr class="odd">
<td align="center">gene_subset_size</td>
<td align="center">number of genes used for a block (used in Reduce-Recover mode).</td>
</tr>
<tr class="even">
<td align="center">block_number</td>
<td align="center">number of blocks (used in Reduce-Recover mode).</td>
</tr>
<tr class="odd">
<td align="center">cpu_number</td>
<td align="center">number of CPUs will be used. If null, CDSeq will detect the number of available cores for parallel computing. Default is NULL.</td>
</tr>
<tr class="even">
<td align="center">gene_length</td>
<td align="center">effective length of the genes in bulk data. It is defined as <em>gene length - read length + 1</em>.</td>
</tr>
<tr class="odd">
<td align="center">reference_gep</td>
<td align="center">pure cell line gene expressions.</td>
</tr>
<tr class="even">
<td align="center">print_progress_msg_to_file</td>
<td align="center">indicator of printing the running process.</td>
</tr>
</tbody>
</table>
</div>
<div id="output" class="section level2">
<h2 class="hasAnchor">
<a href="#output" class="anchor"></a>Ouput</h2>
<table class="table">
<thead><tr class="header">
<th align="left">Output</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">estGEP</td>
<td align="left">CDSeq-estimated cell-type-specific gene expression profiles. It is a G by T matrix where G denotes the number of genes and M is the sample size.</td>
</tr>
<tr class="even">
<td align="left">estProp</td>
<td align="left">CDSeq-estimated sample-specific cell-type proportions. It is a M by T matrix where M is the sample size and T is the number of cell types.</td>
</tr>
<tr class="odd">
<td align="left">cell_type_assignment</td>
<td align="left">If refGEP is given, CDSeq will perform one-to-one cell types assignment for CDSeq-estimated cell types.</td>
</tr>
<tr class="even">
<td align="left">lgpst</td>
<td align="left">This is the log posterior values for each element in cell_type_number.</td>
</tr>
<tr class="odd">
<td align="left">estT</td>
<td align="left">CDSeq-estimated number of cell types.</td>
</tr>
<tr class="even">
<td align="left">est_all</td>
<td align="left">If cell_type_number is a vector, CDSeq will return all estimations for all elements in cell_type_vector.</td>
</tr>
<tr class="odd">
<td align="left">parameters</td>
<td align="left">The user-provided parameters for CDSeq.</td>
</tr>
<tr class="even">
<td align="left">gibbsRunningTime</td>
<td align="left">Time consumed for Gibbs sampler.</td>
</tr>
<tr class="odd">
<td align="left">processIDs</td>
<td align="left">Process ID in the operating system for running CDSeq.</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>You could install CDSeq from Github by running the following</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">install_github</span><span class="op">(</span><span class="st">"kkang7/CDSeq_R_Package"</span><span class="op">)</span></code></pre></div>
</div>
<div id="examples" class="section level1">
<h1 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h1>
<div id="scalar" class="section level2">
<h2 class="hasAnchor">
<a href="#scalar" class="anchor"></a>When the number of cell types is a scalar</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">CDSeq</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/">ggpubr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></code></pre></div>
<p>The following code runs <strong>CDSeq</strong> using synthetic mixtures comes with the R package. The <em>mixtureGEP</em> consists of 40 samples with 500 genes. These <em>in silico</em> mixtures are generated using six pure cell lines and random generated mixing proportions. <strong>CDSeq</strong> takes <em>mixtureGEP</em> as input and outputs the estimates of cell-type-specific gene expressions and sample-specific cell-type proportions. When <em>refGEP</em> is provided, CDSeq will assign the identities of those estimated cell types by comparing the GEP of estimated cell types with the GEP of <em>refGEP</em>. If <em>refGEP</em> is not provided, the estimated cell types will remain anonymous. <a href="#scAnnotate">The other way to annotate CDSeq-estimated cell types is to use single cell RNAseq data</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R">              
<span class="va">result1</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/CDSeq.html">CDSeq</a></span><span class="op">(</span>bulk_data <span class="op">=</span>  <span class="va">mixtureGEP</span>, 
               cell_type_number <span class="op">=</span> <span class="fl">6</span>, 
               mcmc_iterations <span class="op">=</span> <span class="fl">2000</span>, 
               gene_length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="va">gene_length</span><span class="op">)</span>, 
               reference_gep <span class="op">=</span> <span class="va">refGEP</span>,  <span class="co"># gene expression profile of pure cell lines</span>
               cpu_number <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>The output <em>result1</em> is a list containing multiple entities as listed in <a href="#output">Output</a>. When <em>cell_type_number</em> is a scalar, <strong>CDSeq</strong> will only return the following values.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="va">result1</span><span class="op">)</span>
<span class="co">#&gt; [1] "cell_type_assignment" "cellTypeAssignSplit"  "estGEP"              </span>
<span class="co">#&gt; [4] "estProp"              "gibbsRunningTime"     "parameters"          </span>
<span class="co">#&gt; [7] "processIDs"</span></code></pre></div>
<p>We can compare the CDSeq-estimated cell-type-specific GEPs with the ground truth.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trueGEP</span> <span class="op">&lt;-</span> <span class="va">true_GEP_rpkm</span><span class="op">[</span>,<span class="va">result1</span><span class="op">$</span><span class="va">cell_type_assignment</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">4</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span>,mai <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">.1</span>, <span class="fl">.01</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, pty<span class="op">=</span><span class="st">"s"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">max_v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">result1</span><span class="op">$</span><span class="va">estGEP</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>, <span class="va">trueGEP</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="va">result1</span><span class="op">$</span><span class="va">estGEP</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>,
       <span class="va">trueGEP</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>,
       xlab <span class="op">=</span> <span class="st">"CDSeq-estimated GEPs"</span>, 
       ylab <span class="op">=</span> <span class="st">"True GEPs"</span>,
       pch <span class="op">=</span> <span class="fl">21</span>,
       bg <span class="op">=</span> <span class="st">'green'</span>,
       cex<span class="op">=</span><span class="fl">2</span>, 
       xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">max_v</span><span class="op">)</span>, 
       ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="va">max_v</span><span class="op">)</span>,
       main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"cell type "</span>,<span class="va">i</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="va">max_v</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="va">max_v</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="CDSeq-vignette_files/figure-html/fig1-1.png" width="672" style="display: block; margin: auto;"></p>
<p>We then compare the CDSeq-estimated cell type proportions with group truth.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trueProp</span> <span class="op">&lt;-</span> <span class="va">true_prop_cell</span><span class="op">[</span><span class="va">result1</span><span class="op">$</span><span class="va">cell_type_assignment</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">4</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span>,mai <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">.1</span>, <span class="fl">.01</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, pty<span class="op">=</span><span class="st">"s"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="va">result1</span><span class="op">$</span><span class="va">estProp</span><span class="op">[</span><span class="va">i</span>,<span class="op">]</span>,
         <span class="va">trueProp</span><span class="op">[</span><span class="va">i</span>,<span class="op">]</span>,
         pch <span class="op">=</span> <span class="fl">21</span>,
         cex <span class="op">=</span> <span class="fl">3</span>,
         bg<span class="op">=</span><span class="st">'red'</span>,
         xlab <span class="op">=</span> <span class="st">"CDSeq-estimated proportions"</span>, 
         ylab <span class="op">=</span> <span class="st">"True proportions"</span>, 
         xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,
         ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,
         main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"cell type "</span>,<span class="va">i</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="CDSeq-vignette_files/figure-html/fig2-1.png" width="672" style="display: block; margin: auto;"></p>
<p>For this dataset and given parameter settings, CDSeq took this much amount of time to finish:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">result1</span><span class="op">$</span><span class="va">gibbsRunningTime</span>
<span class="co">#&gt;     user   system  elapsed </span>
<span class="co">#&gt; 2292.907    2.416 2295.437</span></code></pre></div>
</div>
<div id="vector" class="section level2">
<h2 class="hasAnchor">
<a href="#vector" class="anchor"></a>When the number of cell types is a vector</h2>
<p>The <em>cell_type_number</em> can also be a vector which contains different integer values. CDSeq will perform estimation for each integer in the vector and estimate the number of cell types in the mixtures. For example, one can set <em>cell_type_number = 2:10</em> as follows, and CDSeq will estimate the most likely number of cell types from 2 to 10.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">result2</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/CDSeq.html">CDSeq</a></span><span class="op">(</span>bulk_data <span class="op">=</span>  <span class="va">mixtureGEP</span>, 
              cell_type_number <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">10</span>, 
              mcmc_iterations <span class="op">=</span> <span class="fl">2000</span>, 
              dilution_factor <span class="op">=</span> <span class="fl">1</span>, 
              block_number <span class="op">=</span> <span class="fl">1</span>, 
              gene_length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="va">gene_length</span><span class="op">)</span>, 
              reference_gep <span class="op">=</span> <span class="va">refGEP</span>, <span class="co"># gene expression profile of pure cell lines</span>
              cpu_number <span class="op">=</span> <span class="fl">9</span>, <span class="co"># use multiple cores to save time. Set the cpu_number = length(cell_type_number) if there is enough cores.</span>
              print_progress_msg_to_file <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></code></pre></div>
<p>CDSeq will automatically estimate the number of cell types from the <em>cell_type_number</em> vector (2:10 in our example) and return the corresponding estimate in <em>estGEP</em> and <em>estProp</em> and it will keep all the estimations for the different values in <em>cell_type_number</em> in <em>est_all</em>. Check out <a href="#output">output</a> for details.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trueGEP</span> <span class="op">&lt;-</span> <span class="va">true_GEP_gene</span><span class="op">[</span>,<span class="va">result2</span><span class="op">$</span><span class="va">cell_type_assignment</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">4</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span>,mai <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">.1</span>, <span class="fl">.01</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, pty<span class="op">=</span><span class="st">"s"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">max_v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">result2</span><span class="op">$</span><span class="va">estGEP</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>, <span class="va">trueGEP</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="va">result2</span><span class="op">$</span><span class="va">estGEP</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>,
       <span class="va">trueGEP</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>,
       xlab <span class="op">=</span> <span class="st">"CDSeq-estimated GEPs"</span>, 
       ylab <span class="op">=</span> <span class="st">"True GEPs"</span>,
       pch <span class="op">=</span> <span class="fl">21</span>,
       bg <span class="op">=</span> <span class="st">'green'</span>,
       cex<span class="op">=</span><span class="fl">2</span>, 
       xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">max_v</span><span class="op">)</span>, 
       ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="va">max_v</span><span class="op">)</span>,
       main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"cell type "</span>,<span class="va">i</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="va">max_v</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="va">max_v</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="CDSeq-vignette_files/figure-html/fig3-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Similarly, we can compare the CDSeq-estimated cell type proportions with ground truth.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trueProp</span> <span class="op">&lt;-</span> <span class="va">true_prop_cell</span><span class="op">[</span><span class="va">result2</span><span class="op">$</span><span class="va">cell_type_assignment</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">4</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span>,mai <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">.1</span>, <span class="fl">.01</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, pty<span class="op">=</span><span class="st">"s"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="va">result2</span><span class="op">$</span><span class="va">estProp</span><span class="op">[</span><span class="va">i</span>,<span class="op">]</span>,
         <span class="va">trueProp</span><span class="op">[</span><span class="va">i</span>,<span class="op">]</span>,
         pch <span class="op">=</span> <span class="fl">21</span>,
         cex <span class="op">=</span> <span class="fl">3</span>,
         bg<span class="op">=</span><span class="st">'red'</span>,
         xlab <span class="op">=</span> <span class="st">"CDSeq-estimated proportions"</span>, 
         ylab <span class="op">=</span> <span class="st">"True proportions"</span>, 
         xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,
         ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,
         main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"cell type "</span>,<span class="va">i</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="CDSeq-vignette_files/figure-html/fig4-1.png" width="672" style="display: block; margin: auto;"></p>
<p>CDSeq estimates the number of cell types in the bulk samples using log posterior values. You can plot the log posterior values as follows. In this example, the max log posterior value happens at 6, therefore, CDSeq will guess that the number of different cell types in the bulk samples is 6 (which is exactly the number of distinct cell types we used for generating this <em>in silico</em> mixtures).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">lgpst</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">9</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">lgpst</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">result2</span><span class="op">$</span><span class="va">est_all</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">lgpst</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">10</span>,<span class="va">lgpst</span>,xlab <span class="op">=</span> <span class="st">"number of cell typess"</span>, ylab <span class="op">=</span> <span class="st">"log posterior"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span><span class="fl">6</span>,<span class="va">lgpst</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span>,pch<span class="op">=</span><span class="fl">16</span>,col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">10</span>, <span class="va">lgpst</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">6</span>,<span class="fl">6</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="va">lgpst</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span>,lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span></code></pre></div>
<p><img src="CDSeq-vignette_files/figure-html/fig5-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div id="use-dilution-factor-to-speed-up-cdseq" class="section level2">
<h2 class="hasAnchor">
<a href="#use-dilution-factor-to-speed-up-cdseq" class="anchor"></a>Use dilution factor to speed up CDSeq</h2>
<p>To speed up CDSeq, you can use <em>dilution_factor</em>. Set <em>dilution_factor</em> equal to <em>N</em>, CDSeq will require <em>N</em> times less computation time. The cost is that using dilution factor will <a href="#input">lose information in the bulk samples</a>. In practice, one can first use a large value for <em>dilution_factor</em> and estimate how much time will be required for running CDSeq. For example, if setting <em>dilution_factor=1000</em> takes CDSeq 1 minute to finish, then setting <em>dilution_factor=1</em> will roughly take CDSeq 1000 minutes to finish. Depending on the available computational power and time, one can adjust the value of <em>dilution_factor</em> to the affordably miminum value.</p>
</div>
<div id="reduce-recover-strategy-for-speeding-up" class="section level2">
<h2 class="hasAnchor">
<a href="#reduce-recover-strategy-for-speeding-up" class="anchor"></a>Reduce-Recover strategy for speeding up</h2>
<p>We provide another option to speed up CDSeq.The idea of Reduce-Recover is to randomly sample multiple subsets of genes and perform deconvolution on these subsets of genes in parallel and combine the estimations from these subsets and take the average as the final estimates.</p>
<p>To use this feature, you need to set values for <em>block_number</em> and <em>gene_subset_size</em>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">result3</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/CDSeq.html">CDSeq</a></span><span class="op">(</span>bulk_data <span class="op">=</span>  <span class="va">mixtureGEP</span>, 
              cell_type_number <span class="op">=</span> <span class="fl">6</span>, 
              mcmc_iterations <span class="op">=</span> <span class="fl">2000</span>, 
              dilution_factor <span class="op">=</span> <span class="fl">1</span>, 
              block_number <span class="op">=</span> <span class="fl">10</span>,
              gene_subset_size <span class="op">=</span> <span class="fl">100</span>,
              gene_length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="va">gene_length</span><span class="op">)</span>, 
              reference_gep <span class="op">=</span> <span class="va">refGEP</span>, 
              cpu_number <span class="op">=</span> <span class="fl">10</span>, 
              print_progress_msg_to_file <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">trueGEP</span> <span class="op">&lt;-</span> <span class="va">true_GEP_rpkm</span><span class="op">[</span>,<span class="va">result3</span><span class="op">$</span><span class="va">cell_type_assignment</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">4</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span>,mai <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">.1</span>, <span class="fl">.01</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, pty<span class="op">=</span><span class="st">"s"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">max_v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">result3</span><span class="op">$</span><span class="va">estGEP</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>, <span class="va">trueGEP</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="va">result3</span><span class="op">$</span><span class="va">estGEP</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>,
       <span class="va">trueGEP</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>,
       xlab <span class="op">=</span> <span class="st">"CDSeq-estimated GEPs"</span>, 
       ylab <span class="op">=</span> <span class="st">"True GEPs"</span>,
       pch <span class="op">=</span> <span class="fl">21</span>,
       bg <span class="op">=</span> <span class="st">'green'</span>,
       cex<span class="op">=</span><span class="fl">2</span>, 
       xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">max_v</span><span class="op">)</span>, 
       ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="va">max_v</span><span class="op">)</span>,
       main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"cell type "</span>,<span class="va">i</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="va">max_v</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="va">max_v</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="CDSeq-vignette_files/figure-html/fig6-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">trueProp</span> <span class="op">&lt;-</span> <span class="va">true_prop_cell</span><span class="op">[</span><span class="va">result3</span><span class="op">$</span><span class="va">cell_type_assignment</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">4</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span>,mai <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">.1</span>, <span class="fl">.01</span>, <span class="fl">0.1</span>, <span class="fl">0.1</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, pty<span class="op">=</span><span class="st">"s"</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span><span class="op">(</span><span class="va">result3</span><span class="op">$</span><span class="va">estProp</span><span class="op">[</span><span class="va">i</span>,<span class="op">]</span>,
         <span class="va">trueProp</span><span class="op">[</span><span class="va">i</span>,<span class="op">]</span>,
         pch <span class="op">=</span> <span class="fl">21</span>,
         cex <span class="op">=</span> <span class="fl">3</span>,
         bg<span class="op">=</span><span class="st">'red'</span>,
         xlab <span class="op">=</span> <span class="st">"CDSeq-estimated proportions"</span>, 
         ylab <span class="op">=</span> <span class="st">"True proportions"</span>, 
         xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,
         ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,
         main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"cell type "</span>,<span class="va">i</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"l"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p><img src="CDSeq-vignette_files/figure-html/fig7-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="scAnnotate" class="section level1">
<h1 class="hasAnchor">
<a href="#scAnnotate" class="anchor"></a>Annotate CDSeq-estimated cell types using scRNASeq</h1>
<p>Another way to annotate CDSeq-estimated cell types is to use single cell RNAseq data with annotations. The idea is to combine the GEPs of CDSeq-estimated cell types with scRNAseq GEPs and perform clustering. Based on the clustering result and the known scRNAseq annotation, we can annotate each cluster to the cell type to which the majority of the single cells in the cluster belong; then assigning the CDSeq-identified cell types accordingly. CDSeq employs functions in Seurat package to perform the clustering analysis.</p>
<p>We demonstrate this function using a set of 30 PBMC mixtures which is constructed using PBMC scRNAseq data. We set the <em>cell_type_number = c(3,6,9,12)</em> and investigate what the impacts of using different number of cell types could be.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cdseq.result</span> <span class="op">&lt;-</span> <span class="fu">CDSeq</span><span class="fu">::</span><span class="fu"><a href="../reference/CDSeq.html">CDSeq</a></span><span class="op">(</span>bulk_data <span class="op">=</span> <span class="va">pbmc_mix</span>,
                             cell_type_number <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">12</span>,<span class="fl">3</span><span class="op">)</span>,
                             beta <span class="op">=</span> <span class="fl">0.5</span>,
                             alpha <span class="op">=</span> <span class="fl">5</span>,
                             mcmc_iterations <span class="op">=</span> <span class="fl">700</span>,
                             cpu_number <span class="op">=</span> <span class="fl">4</span>,
                             dilution_factor <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p>We then use <em>cellTypeAssignSCRNA</em> function to annotate CDSeq-estimated cell types for all the outputs.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cdseq.result.celltypeassign</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, length <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>

<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">cdseq_gep</span> <span class="op">&lt;-</span> <span class="va">cdseq.result</span><span class="op">$</span><span class="va">est_all</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">estGEP</span>
  <span class="va">cdseq_prop</span> <span class="op">&lt;-</span> <span class="va">cdseq.result</span><span class="op">$</span><span class="va">est_all</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">estProp</span>
  <span class="va">cdseq.result.celltypeassign</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cellTypeAssignSCRNA.html">cellTypeAssignSCRNA</a></span><span class="op">(</span>cdseq_gep <span class="op">=</span> <span class="va">cdseq_gep</span>, <span class="co"># CDSeq-estimated cell-type-specific GEPs</span>
                                                     cdseq_prop <span class="op">=</span> <span class="va">cdseq_prop</span>, <span class="co"># CDSeq-estimated cell type proportions</span>
                                                     sc_gep <span class="op">=</span> <span class="va">sc_gep</span>,         <span class="co"># PBMC single cell data</span>
                                                     sc_annotation <span class="op">=</span> <span class="va">sc_annotation</span>,<span class="co"># PBMC single data annotations</span>
                                                     sc_pt_size <span class="op">=</span> <span class="fl">3</span>,
                                                     cdseq_pt_size <span class="op">=</span> <span class="fl">6</span>,
                                                     seurat_nfeatures <span class="op">=</span> <span class="fl">100</span>,
                                                     seurat_npcs <span class="op">=</span> <span class="fl">50</span>,
                                                     seurat_dims<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,
                                                     plot_umap <span class="op">=</span> <span class="fl">1</span>,
                                                     plot_tsne <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
  
  <span class="va">cdseq.result.celltypeassign</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">cdseq_scRNA_umap</span>
<span class="op">}</span></code></pre></div>
<p>We now show the umap 2D embeddings of the CDSeq-estimated cell-type-specific GEPs and scRNAseq GEPs. In the following figures, the dots represent PBMC scRNAseq and the squares represent the CDSeq-estimated cell types. The different colors represent the distinct cell types.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span><span class="va">cdseq.result.celltypeassign</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">cdseq_scRNA_umap</span>, 
          <span class="va">cdseq.result.celltypeassign</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">cdseq_scRNA_umap</span>,
          <span class="va">cdseq.result.celltypeassign</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">cdseq_scRNA_umap</span>,
          <span class="va">cdseq.result.celltypeassign</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">cdseq_scRNA_umap</span>,
          nrow <span class="op">=</span> <span class="fl">4</span>,ncol <span class="op">=</span> <span class="fl">1</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </code></pre></div>
<p><img src="CDSeq-vignette_files/figure-html/show%20GEP%20vs%20scRNAseq-1.png" width="95%" style="display: block; margin: auto;"></p>
<p>We next compare the CDSeq-estimated sample-specific cell-type proportions with the ground truth. We first prepare data frame for ggplot.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>) {</span>
<span id="cb19-2"><a href="#cb19-2"></a>  <span class="cf">if</span>(i<span class="op">==</span><span class="dv">1</span>){</span>
<span id="cb19-3"><a href="#cb19-3"></a>    <span class="kw">rownames</span>(cdseq.result.celltypeassign[[i]]<span class="op">$</span>cdseq_prop_merged) &lt;-<span class="st"> </span><span class="kw">rownames</span>(true_prop)</span>
<span id="cb19-4"><a href="#cb19-4"></a>    cdseq_prop_df &lt;-<span class="st"> </span><span class="kw">melt</span>(cdseq.result.celltypeassign[[i]]<span class="op">$</span>cdseq_prop_merged)</span>
<span id="cb19-5"><a href="#cb19-5"></a>    <span class="kw">names</span>(cdseq_prop_df)[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"celltype"</span>, <span class="st">"sampleID"</span>,<span class="kw">paste0</span>(<span class="st">"estprop_"</span>,i))</span>
<span id="cb19-6"><a href="#cb19-6"></a>    true_prop_df &lt;-<span class="st"> </span><span class="kw">melt</span>(true_prop)</span>
<span id="cb19-7"><a href="#cb19-7"></a>    <span class="kw">names</span>(true_prop_df)[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"celltype"</span>, <span class="st">"sampleID"</span>,<span class="st">"trueprop"</span>)</span>
<span id="cb19-8"><a href="#cb19-8"></a>    </span>
<span id="cb19-9"><a href="#cb19-9"></a>    merge_df &lt;-<span class="st"> </span><span class="kw">merge</span>(cdseq_prop_df,true_prop_df, <span class="dt">by=</span><span class="kw">c</span>(<span class="st">"celltype"</span>,<span class="st">"sampleID"</span>))</span>
<span id="cb19-10"><a href="#cb19-10"></a>  }<span class="cf">else</span>{</span>
<span id="cb19-11"><a href="#cb19-11"></a>    <span class="kw">rownames</span>(cdseq.result.celltypeassign[[i]]<span class="op">$</span>cdseq_prop_merged) &lt;-<span class="st"> </span><span class="kw">rownames</span>(true_prop)</span>
<span id="cb19-12"><a href="#cb19-12"></a>    cdseq_prop_df &lt;-<span class="st"> </span><span class="kw">melt</span>(cdseq.result.celltypeassign[[i]]<span class="op">$</span>cdseq_prop_merged)</span>
<span id="cb19-13"><a href="#cb19-13"></a>    <span class="kw">names</span>(cdseq_prop_df)[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"celltype"</span>, <span class="st">"sampleID"</span>,<span class="kw">paste0</span>(<span class="st">"estprop_"</span>,i))</span>
<span id="cb19-14"><a href="#cb19-14"></a>  </span>
<span id="cb19-15"><a href="#cb19-15"></a>    merge_df &lt;-<span class="st"> </span><span class="kw">merge</span>(merge_df, cdseq_prop_df, <span class="dt">by=</span><span class="kw">c</span>(<span class="st">"celltype"</span>,<span class="st">"sampleID"</span>))</span>
<span id="cb19-16"><a href="#cb19-16"></a>  }</span>
<span id="cb19-17"><a href="#cb19-17"></a>}</span>
<span id="cb19-18"><a href="#cb19-18"></a><span class="er">}</span></span></code></pre></div>
<p>Then, we can compare CDSeq estimated proportion with ground truth. Note that CDSeq will merge the estimated cell type proportion of the same type after cell type annotation.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc_ggplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, length <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">B_plot</span><span class="op">&lt;-</span> <span class="va">merge_df</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">celltype</span><span class="op">==</span><span class="st">"Bcell"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggscatter.html">ggscatter</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"estprop_"</span>,<span class="va">i</span><span class="op">)</span>, y <span class="op">=</span> <span class="st">"trueprop"</span>,
                                                                add <span class="op">=</span> <span class="st">"reg.line"</span>,  
                                                                size <span class="op">=</span> <span class="fl">3</span>,
                                                                title <span class="op">=</span> <span class="st">"B cell"</span>,
                                                                add.params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"blue"</span>, fill <span class="op">=</span> <span class="st">"lightgray"</span><span class="op">)</span>, <span class="co"># Customize reg. line</span>
                                                                conf.int <span class="op">=</span> <span class="cn">TRUE</span> 
  <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/stat_cor.html">stat_cor</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"pearson"</span>, p.accuracy <span class="op">=</span> <span class="fl">0.001</span>, r.accuracy <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>
  
  <span class="va">T_plot</span><span class="op">&lt;-</span> <span class="va">merge_df</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">celltype</span><span class="op">==</span><span class="st">"Tcell"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggscatter.html">ggscatter</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"estprop_"</span>,<span class="va">i</span><span class="op">)</span>, y <span class="op">=</span> <span class="st">"trueprop"</span>,
                                                                add <span class="op">=</span> <span class="st">"reg.line"</span>, 
                                                                size <span class="op">=</span> <span class="fl">3</span>,
                                                                title <span class="op">=</span> <span class="st">"T cell"</span>,
                                                                add.params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"blue"</span>, fill <span class="op">=</span> <span class="st">"lightgray"</span><span class="op">)</span>, <span class="co"># Customize reg. line</span>
                                                                conf.int <span class="op">=</span> <span class="cn">TRUE</span> <span class="co"># Add confidence interval</span>
  <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/stat_cor.html">stat_cor</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"pearson"</span>, p.accuracy <span class="op">=</span> <span class="fl">0.001</span>, r.accuracy <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>
  
  <span class="va">M_plot</span><span class="op">&lt;-</span> <span class="va">merge_df</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">celltype</span><span class="op">==</span><span class="st">"Monocyte"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggscatter.html">ggscatter</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"estprop_"</span>,<span class="va">i</span><span class="op">)</span>, y <span class="op">=</span> <span class="st">"trueprop"</span>,
                                                                   add <span class="op">=</span> <span class="st">"reg.line"</span>,  
                                                                   size <span class="op">=</span> <span class="fl">3</span>,
                                                                   title <span class="op">=</span> <span class="st">"Monocyte"</span>,
                                                                   add.params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"blue"</span>, fill <span class="op">=</span> <span class="st">"lightgray"</span><span class="op">)</span>, <span class="co"># Customize reg. line</span>
                                                                   conf.int <span class="op">=</span> <span class="cn">TRUE</span> 
  <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/stat_cor.html">stat_cor</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"pearson"</span>, p.accuracy <span class="op">=</span> <span class="fl">0.001</span>, r.accuracy <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>
  
  
  <span class="va">pbmc_ggplot</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span><span class="va">B_plot</span>, <span class="va">T_plot</span>, <span class="va">M_plot</span>, ncol <span class="op">=</span> <span class="fl">3</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>The first row of the following figures represent the CDSeq results for 3 cell types. The second row of the following figures represent the CDSeq results for 6 cell types. The third row of the following figures represent the CDSeq results for 9 cell types. The last row of the following figures represent the CDSeq results for 12 cell types.</p>
<p>These results suggest that CDSeq provides robust cell-type-specific GEP estimations and sample-specific cell-type proportion estimations when using different number of cell types.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span><span class="va">pbmc_ggplot</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, 
          <span class="va">pbmc_ggplot</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, 
          <span class="va">pbmc_ggplot</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>, 
          <span class="va">pbmc_ggplot</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span>,
          nrow <span class="op">=</span> <span class="fl">4</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="CDSeq-vignette_files/figure-html/show%20plots-1.png" width="95%" style="display: block; margin: auto;"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Kai Kang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
